File: frontend/src/lib/agent-file-roles.ts
export const INSTRUCTION_FILE_PREFIX = "instruction__";
export const OUTPUT_SAMPLE_PREFIX = "output-sample__";
export const INPUT_SAMPLE_PREFIX = "input-sample__";

export type AgentFileRole =
  | "instruction"
  | "output_sample"
  | "input_sample"
  | "runtime";

export function isOutputSampleFileName(name: string, role?: string | null): boolean {
  if (role === "output_sample") return true;
  return name.startsWith(OUTPUT_SAMPLE_PREFIX) || name.includes(OUTPUT_SAMPLE_PREFIX);
}

export function isInstructionFileName(name: string, role?: string | null): boolean {
  if (role === "instruction") return true;
  return name.startsWith(INSTRUCTION_FILE_PREFIX) || name.includes(INSTRUCTION_FILE_PREFIX);
}

export function isInputSampleFileName(name: string, role?: string | null): boolean {
  if (role === "input_sample") return true;
  return name.startsWith(INPUT_SAMPLE_PREFIX) || name.includes(INPUT_SAMPLE_PREFIX);
}

/** Unprefixed runtime / training inputs (not instruction or output-sample). */
export function isRuntimeSampleFileName(name: string, role?: string | null): boolean {
  if (role === "runtime" || role === "input_sample") return true;
  if (role === "instruction" || role === "output_sample") return false;
  return !isInstructionFileName(name) && !isOutputSampleFileName(name);
}

/** Training / verify input samples: input-sample__ or unprefixed runtime files. */
export function isTrainingInputFileName(name: string, role?: string | null): boolean {
  return isRuntimeSampleFileName(name, role);
}

export function agentFileRoleFromName(name: string, role?: string | null): AgentFileRole {
  if (role === "instruction" || role === "output_sample" || role === "input_sample" || role === "runtime") {
    return role;
  }
  if (isInstructionFileName(name)) return "instruction";
  if (isOutputSampleFileName(name)) return "output_sample";
  if (isInputSampleFileName(name)) return "input_sample";
  return "runtime";
}

/** Server-side agent file helpers (use role column when present). */
export function isServerInstructionFile(file: { filename: string; role?: string | null }): boolean {
  return isInstructionFileName(file.filename, file.role);
}

export function isServerOutputSampleFile(file: { filename: string; role?: string | null }): boolean {
  return isOutputSampleFileName(file.filename, file.role);
}

export function isServerInputSampleFile(file: { filename: string; role?: string | null }): boolean {
  return isRuntimeSampleFileName(file.filename, file.role);
}

export function displayAgentFileName(name: string): string {
  for (const prefix of [INSTRUCTION_FILE_PREFIX, OUTPUT_SAMPLE_PREFIX, INPUT_SAMPLE_PREFIX]) {
    if (name.startsWith(prefix)) return name.slice(prefix.length);
    const i = name.indexOf(prefix);
    if (i >= 0) return name.slice(i + prefix.length);
  }
  return name;
}

export function asInstructionFile(file: File): File {
  if (isInstructionFileName(file.name)) return file;
  return new File([file], `${INSTRUCTION_FILE_PREFIX}${file.name}`, {
    type: file.type,
    lastModified: file.lastModified,
  });
}

export function asOutputSampleFile(file: File): File {
  if (isOutputSampleFileName(file.name)) return file;
  return new File([file], `${OUTPUT_SAMPLE_PREFIX}${file.name}`, {
    type: file.type,
    lastModified: file.lastModified,
  });
}

export function asInputSampleFile(file: File): File {
  if (isInputSampleFileName(file.name)) return file;
  if (isInstructionFileName(file.name) || isOutputSampleFileName(file.name)) return file;
  return new File([file], `${INPUT_SAMPLE_PREFIX}${file.name}`, {
    type: file.type,
    lastModified: file.lastModified,
  });
}

export function roleBadgeLabel(role: AgentFileRole): string {
  switch (role) {
    case "instruction":
      return "دستورالعمل";
    case "output_sample":
      return "نمونه خروجی";
    case "input_sample":
      return "نمونه ورودی";
    default:
      return "فایل";
  }
}

File: backend/src/core/agent_file_roles.py
"""Filename-role helpers for agent attachments.

Roles may also live on ``AgentFile.role`` (nullable DB column). Filename
prefixes remain the backward-compatible source of truth for uploads that
predate the column.
"""

from __future__ import annotations

INSTRUCTION_FILE_PREFIX = "instruction__"
OUTPUT_SAMPLE_PREFIX = "output-sample__"
INPUT_SAMPLE_PREFIX = "input-sample__"

# Canonical role strings stored on AgentFile.role / config / meta.
ROLE_INSTRUCTION = "instruction"
ROLE_OUTPUT_SAMPLE = "output_sample"
ROLE_INPUT_SAMPLE = "input_sample"
ROLE_RUNTIME = "runtime"

VALID_ROLES = frozenset(
    {ROLE_INSTRUCTION, ROLE_OUTPUT_SAMPLE, ROLE_INPUT_SAMPLE, ROLE_RUNTIME}
)


def is_instruction_file(filename: str | None) -> bool:
    name = filename or ""
    return name.startswith(INSTRUCTION_FILE_PREFIX) or INSTRUCTION_FILE_PREFIX in name


def is_output_sample_file(filename: str | None) -> bool:
    name = filename or ""
    return name.startswith(OUTPUT_SAMPLE_PREFIX) or OUTPUT_SAMPLE_PREFIX in name


def is_input_sample_file(filename: str | None) -> bool:
    name = filename or ""
    return name.startswith(INPUT_SAMPLE_PREFIX) or INPUT_SAMPLE_PREFIX in name


def agent_file_role(filename: str | None, *, role: str | None = None) -> str:
    """Resolve role: explicit DB role wins when valid, else filename prefixes."""
    if role and role in VALID_ROLES:
        return role
    if is_instruction_file(filename):
        return ROLE_INSTRUCTION
    if is_output_sample_file(filename):
        return ROLE_OUTPUT_SAMPLE
    if is_input_sample_file(filename):
        return ROLE_INPUT_SAMPLE
    return ROLE_RUNTIME


def is_training_input_sample(filename: str | None, *, role: str | None = None) -> bool:
    """True for files used as verify/synth *input* samples (not instruction/output)."""
    r = agent_file_role(filename, role=role)
    return r in {ROLE_INPUT_SAMPLE, ROLE_RUNTIME} and r != ROLE_INSTRUCTION


def display_agent_filename(filename: str | None) -> str:
    name = filename or ""
    for prefix in (INSTRUCTION_FILE_PREFIX, OUTPUT_SAMPLE_PREFIX, INPUT_SAMPLE_PREFIX):
        if name.startswith(prefix):
            return name[len(prefix) :]
        # uuid_prefix__name storage form
        if prefix in name:
            idx = name.find(prefix)
            if idx >= 0:
                return name[idx + len(prefix) :]
    return name


def pair_id_from_filename(filename: str | None) -> str | None:
    """Optional pair tag: ``pairN__`` segment after role prefix, or ``pairN`` in stem."""
    import re

    name = display_agent_filename(filename)
    m = re.search(r"(?:^|[_\-])pair([a-zA-Z0-9]+)(?:[_\-]|$)", name, flags=re.I)
    if m:
        return f"pair{m.group(1)}"
    m2 = re.match(r"pair([a-zA-Z0-9]+)[_\-]", name, flags=re.I)
    if m2:
        return f"pair{m2.group(1)}"
    return None
